### PR TITLE
add restartall API to the flow service

### PR DESCRIPTION
**Pull request by msandg**

### CODE REVIEW

**Comment by msandg** [comment]
Jose's comment:

I like this implementation. I want to confirm with @rohit489 , is this new endpoint a test candidate for this suite of tests => https://github.com/microsoft/data-accelerator/tree/master/Tests/DataXScenarios/DataXScenarios ? If thats the case, i think it would be good to add a test call to this new endpoint there

File: Services/DataX.Flow/Flow.ManagementService/Controllers/FlowManagementController.cs
Comment (rohit489) [inline Services/DataX.Flow/Flow.ManagementService/Controllers/FlowManagementController.cs:425]: Should we make it HttpPost. Will also help if take take an input list of jobnames to restart

File: Services/DataX.Config/DataX.Config/PublicService/JobOperation.cs
Comment (rohit489) [inline Services/DataX.Config/DataX.Config/PublicService/JobOperation.cs:58]: We should add a check to verify if the job was restarted fine. If we restart a lot of jobs in a small time interval we have seen failures in job restart where HDInsight fails to allocate resources to the jobs

Comment (rohit489) [reply Services/DataX.Config/DataX.Config/PublicService/JobOperation.cs:58]: We should also add a retry logic in case of failures

Comment (rohit489) [reply Services/DataX.Config/DataX.Config/PublicService/JobOperation.cs:58]: As we restart a batch of jobs in a given list of jobs, we can call SyncJobStateByNames() to check the status of each job in the batch and once all the jobs are in running state then proceed to next batch in the list. If some job was not in running state after a threshoold time period then we can issue a restart again for that job

**Comment by rohit489** [comment]
> Jose's comment:
> 
> I like this implementation. I want to confirm with @rohit489 , is this new endpoint a test candidate for this suite of tests => https://github.com/microsoft/data-accelerator/tree/master/Tests/DataXScenarios/DataXScenarios ? If thats the case, i think it would be good to add a test call to this new endpoint there

Yes makes sense to add this to scenario tester as well like @jozavala-msft 

> Jose's comment:
> 
> I like this implementation. I want to confirm with @rohit489 , is this new endpoint a test candidate for this suite of tests => https://github.com/microsoft/data-accelerator/tree/master/Tests/DataXScenarios/DataXScenarios ? If thats the case, i think it would be good to add a test call to this new endpoint there

Yes makes sense to add it to scenario tester like @jozavala-msft  pointed

File: Services/DataX.Config/DataX.Config/PublicService/JobOperation.cs
Comment (msandg) [reply Services/DataX.Config/DataX.Config/PublicService/JobOperation.cs]: 1. added a new API to wait for jobs up
2. added retry logics in the wait API

File: Services/DataX.Flow/Flow.ManagementService/Controllers/FlowManagementController.cs
Comment (rohit489) [inline Services/DataX.Flow/Flow.ManagementService/Controllers/FlowManagementController.cs:504]: In case resources donot get allocated to the spark job on HDInsight, the state changes to idle. In this case job submission needs to be retired too

File: Services/DataX.Flow/Flow.ManagementService/Controllers/FlowManagementController.cs
Comment (rohit489) [inline Services/DataX.Flow/Flow.ManagementService/Controllers/FlowManagementController.cs:505]: in case the job is not running, shouldn't we issue a restart again for that job?

**Review by jozavala-msft — APPROVED** [review]
We agreed that manual testing is enough for this feature. We are covering automated testing in another PR, and we don't need to add scenario tester code here. LGTM. Lets wait for @rohit489 approval as well.

File: Services/DataX.Flow/Flow.ManagementService/Controllers/FlowManagementController.cs
Comment (guiyang2011) [reply Services/DataX.Flow/Flow.ManagementService/Controllers/FlowManagementController.cs]: yes, it would restart when job not running

File: Services/DataX.Flow/Flow.ManagementService/Controllers/FlowManagementController.cs
Comment (msandg) [reply Services/DataX.Flow/Flow.ManagementService/Controllers/FlowManagementController.cs]: have updated the logic to restart the job state is idle, starting or failed

File: Services/DataX.Flow/Flow.ManagementService/Controllers/FlowManagementController.cs
Comment (msandg) [reply Services/DataX.Flow/Flow.ManagementService/Controllers/FlowManagementController.cs]: yes, will restart it

File: Services/DataX.Flow/Flow.ManagementService/Controllers/FlowManagementController.cs
Comment (rohit489) [inline Services/DataX.Flow/Flow.ManagementService/Controllers/FlowManagementController.cs:428]: Instead of taking parameter of type RestatAllWithRetriesRequest, lets use JObject. See "public async Task<ApiResult> SaveFlow([FromBody]JObject config)" for reference

File: Services/DataX.Flow/Flow.ManagementService/Controllers/RestartAllWithRetriesRequest.cs
Comment (rohit489) [inline Services/DataX.Flow/Flow.ManagementService/Controllers/RestartAllWithRetriesRequest.cs:1]: Instead of adding model class in controller, add it in DataX.Config.ConfigDataModel

**Review by rohit489 — APPROVED** [review]

**Pull request merged**
